Warn about NSCoding classes with unstable mangled names

DougGregor opened the pull request:
The name mangling changed from Swift 3 to Swift 4, and may get slight
tweaks as we lock down ABI stability. Identify and warn about (in
Swift 3) or error about (in Swift 4) the cases where we don't have
obviously-stable name mangling, e.g.,

* `private`/`fileprivate` classes (whose mangled names involve the file name)
* nested classes (whose mangled names depend on their enclosing type)
* generic classes (whose mangled names involve the type arguments)

For the first two bullets, one can suppress the warning by applying the newly-introduced `NSKeyedArchiveLegacy` attribute and providing the name of the class that was used in existing archives (i.e., the Swift 3 mangled name).

DougGregor committed:
[Type checker] Warn about NSCoding-conforming classes with unstable mangled names.

The name mangling changed from Swift 3 to Swift 4, and may get slight
tweaks as we lock down ABI stability. Identify and warn about (in
Swift 3) or error about (in Swift 4) the cases where we don't have
obviously-stable name mangling, e.g.,

* private/fileprivate classes (whose mangled names involve the file name)
* nested classes (whose mangled names depend on their enclosing type)
* generic classes (whose mangled names involve the type arguments)

DougGregor committed:
Add the @NSKeyedArchiveLegacy attribute.

This attribute allows one to provide the "legacy" name of a class for
the purposes of archival (via NSCoding). At the moment, it is only
useful for suppressing the warnings/errors about classes with unstable
archiving names.

DougGregor committed:
Add Fix-Its suggesting @NSKeyedArchiveLegacy when it can be used.

DougGregor committed:
Diagnose attempts to apply @NSKeyedArchiveLegacy to a generic class.

Generic classes don't have a single name to register, so disallow
@NSKeyedArchiveLegacy.

DougGregor commented:
@swift-ci please smoke test

jrose-apple commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -659,7 +666,7 @@ void NormalProtocolConformance::setWitness(ValueDecl *requirement,
   assert(getProtocol() == cast<ProtocolDecl>(requirement->getDeclContext()) &&
          "requirement in wrong protocol");
-  assert(Mapping.count(requirement) == 0 && "Witness already known");
+  assert(!hasWitness(requirement) && "Witness already known");
```
Are these sneaking in from a different patch?

DougGregor replied:
Ah, it was a drive-by fix. I'll separate it.

jrose-apple commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -5866,6 +5883,50 @@ void TypeChecker::checkConformancesInContext(DeclContext *dc,
+        Optional<unsigned> kind;
+        if (classDecl->getGenericSignature()) {
+          kind = 3;
+        } else if (classDecl->getDeclContext()->isTypeContext()) {
```
Also applies to local classes, right? How about `isModuleScopeContext`?

DougGregor replied:
Yeah, that'll work better. Thanks!

jrose-apple commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -5866,6 +5883,50 @@ void TypeChecker::checkConformancesInContext(DeclContext *dc,
+        } else if (classDecl->getDeclContext()->isTypeContext()) {
+          kind = 2;
+        } else {
+          switch (classDecl->getFormalAccess()) {
```
This should probably use getFormalAccessScope instead, even though getFormalAccess will give the same answer for top-level classes.

DougGregor replied:
That seems... awkward. `AccessScope` is great when we're trying to determine accessibility of some entity from some specific context. We're asking a simpler question about the stated access level.

jrose-apple replied:
That simpler question would be incorrect for nested clasess, though. The test here isn't about the class's access as stated; it's about whether or not it will have a private discriminator in its runtime name.

jrose-apple commented on test/Interpreter/SDK/archiving_generic_swift_class.swift:
```diff
@@ -1,4 +1,4 @@
-// RUN: %target-run-simple-swift | %FileCheck %s
+// RUN: %target-run-simple-swift -swift-version 3 | %FileCheck %s
```
This isn't actually using Swift 3, which is why you had to update the command line parsing.

DougGregor replied:
Not sure what happened here; I don't need any of this.

jrose-apple commented on include/swift/AST/Attr.def:
```diff
@@ -272,6 +272,8 @@ DECL_ATTR(_implements, Implements,
           | NotSerialized | UserInaccessible,
           /* Not serialized */ 67)
 
+DECL_ATTR(NSKeyedArchiveLegacy, NSKeyedArchiveLegacy, OnClass, 68)
```
This one should be `NotSerialized` too. (I realize that rules out adding the attribute on the class and the conformance in another module, but that's okay.)

Also `LongAttribute`, to improve printing.

DougGregor replied:
I fretted over this. `NotSerialized` means that we won't necessarily see it in the generated interface for the module. While not wrong per se... it seems like useful information.

jrose-apple replied:
I definitely don't want to pollute the generated interface with mangled names, and I don't think it's worth us coming up with a way to print it without printing the mangled name.

DougGregor replied:
I'm fine with dropping it from serialization. Less code to break

jrose-apple commented on test/decl/protocol/conforms/nscoding.swift:
```diff
@@ -0,0 +1,74 @@
+import Foundation
+
+// Top-level classes
+class CodingA : NSObject, NSCoding {
```
Please add a test case for a subclass of a class that already conforms to NSCoding.

DougGregor replied:
It's in the "// Inherited conformances." section below

jrose-apple replied:
Oops, thanks!

jrose-apple commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -5925,6 +5927,12 @@ void TypeChecker::checkConformancesInContext(DeclContext *dc,
+            diagnose(classDecl, diag::add_nskeyedarchivelegacy_attr)
+              .fixItInsert(
+                classDecl->getAttributeInsertionLoc(/*forModifier=*/false),
+                "@NSKeyedArchiveLegacy(\"<#class archival name#>\")");
```
We need to actually provide the old mangled name here, rather than trust users to get it right.

DougGregor replied:
Do we have that capability?

DougGregor replied:
@eeckstein , do we have API for "get the Swift 3 mangled name of this class"?

slavapestov replied:
I don't think so, but if you only care about non-generic classes I would extract the right bits from the old mangler. It should not amount to much.

DougGregor replied:
Yeah, it's only the non-generic cases that we can actually fix.

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1216,7 +1216,9 @@ WARNING(nscoding_unstable_mangled_name_warn,none,
-
+NOTE(add_nskeyedarchivelegacy_attr,none,
+     "add the '@NSKeyedArchiveLegacy' attribute to specify the class name "
+     "used for archiving", ())
```
Can you change this to make this more discouraged? The recommended fix is to use `@objc` (which deserves its own note).

- `"for new classes, add '@objc' to specify a unique, prefixed Objective-C runtime name"`
- `"for compatibility with existing archives, use '@NSKeyedArchiveLegacy' to record the Swift 3 mangled name` (or "internal name", or just "name")

Also, nitpick: I think our diagnostics usually either say `'@foo'` or `'foo' attribute` but not both.

DougGregor replied:
Sure, WFM.

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1219,6 +1219,9 @@ WARNING(nscoding_unstable_mangled_name_warn,none,
      "add the '@NSKeyedArchiveLegacy' attribute to specify the class name "
      "used for archiving", ())
+ERROR(attr_nskeyedarchivelegacy_generic,none,
+      "@NSKeyedArchiveLegacy attribute cannot be applied to generic class %0",
```
Same diagnostic style comment here.

jrose-apple commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -1942,6 +1942,19 @@ void AttributeChecker::visitImplementsAttr(ImplementsAttr *attr) {
+  if (classDecl->getGenericSignatureOfContext()) {
+    TC.diagnose(attr->getLocation(), diag::attr_nskeyedarchivelegacy_generic,
+                classDecl->getDeclaredInterfaceType());
+    attr->setInvalid();
```
Should we use fixItAndRemove here?

DougGregor replied:
`diagnoseAndRemoveAttr` is the thing; I can make it available here

jrose-apple commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -1942,6 +1942,19 @@ void AttributeChecker::visitImplementsAttr(ImplementsAttr *attr) {
 }
 
+void AttributeChecker::visitNSKeyedArchiveLegacyAttr(
+                                              NSKeyedArchiveLegacyAttr *attr) {
```
Nitpick: please double-indent instead of max-indenting.

jrose-apple reviewed (commented):
Thank you for taking this! Many comments so far.

DougGregor committed:
Add the @_staticInitializeObjCMetadata attribute.

Currently inactive, this attribute indicates that a static initializer should be emitted to register the Objective-C metadata when the image is loaded, rather than on first use of the Objective-C metadata. Infer this attribute for NSCoding classes that won’t have static Objective-C metadata or have an @NSKeyedArchiveLegacy attributed.

jrose-apple commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -5934,6 +5947,19 @@ void TypeChecker::checkConformancesInContext(DeclContext *dc,
+        if (!classDecl->getAttrs()
+              .hasAttribute<StaticInitializeObjCMetadataAttr>() &&
+            (kind || hasGenericAncestry(classDecl) ||
+             classDecl->getAttrs().hasAttribute<NSKeyedArchiveLegacyAttr>())) {
```
Doesn't NSKeyedArchiveLegacy mean _staticInitializeObjCMetadata is unnecessary? I guess it doesn't harm anything, but…

DougGregor replied:
Hrm, sure. `IRGen` will end up realizing the ObjC metadata for `@NSKeyedArchiveLegacy` classes

jrose-apple commented on include/swift/AST/Attr.def:
```diff
@@ -274,6 +274,9 @@ DECL_ATTR(_implements, Implements,
 DECL_ATTR(NSKeyedArchiveLegacy, NSKeyedArchiveLegacy, OnClass, 68)
 
+SIMPLE_DECL_ATTR(_staticInitializeObjCMetadata, StaticInitializeObjCMetadata,
+                 OnClass, 69)
```
This one should be UserInaccessible to keep it out of code completion, and should also be NotSerialized.

DougGregor replied:
Sure.

jrose-apple commented on lib/AST/ASTDumper.cpp:
```diff
@@ -855,6 +855,8 @@ namespace {
     void visitClassDecl(ClassDecl *CD) {
       printCommon(CD, "class_decl");
+      if (CD->getAttrs().hasAttribute<StaticInitializeObjCMetadataAttr>())
+        OS << " @_staticInitializeObjCMetadata";
```
Do we not dump attributes today? Ugh.

DougGregor replied:
Nope, we dump specific ones. This yak can remain hairy for today

DougGregor committed:
Properly warn about local classes that conform to NSCoding.

DougGregor committed:
Make diagnostics for NSCoding classes with unstable names more discouraging.

The diagnostic regarding NSCoding classes with unstable names can be
suppressed by adding @objc (the preferred solution for new code) or
@NSKeyedArchiveLegacy (for existing archives). Provide those as
Fix-Its, in that order.

(Thanks, Jordan!)

slavapestov commented:
Do classes in local context need any special handling?

DougGregor committed:
Add @NSKeyedArchiveSubclassesOnly to suppress NSCoding unstable name diags.

Introduce the @NSKeyedArchiveSubclassesOnly attribute, which can be
placed on a class that conforms to NSCoding to suppress the
unstable-name diagnostics by promising to only archive
subclasses---not this class directly.

DougGregor commented:
I'll handle local classes in the update that's en route

DougGregor committed:
Improve/clarify the inference of @_staticInitializeObjCMetadata.

Infer @_staticInitializeObjCMetadata in those cases where we need a
static initializer to make an NSCoding-conforming class visible to the
Objective-C runtime. This does *not* include classes with one of the
@NSKeyedArchive attributes:

* @NSKeyedArchiveLegacy implies that we'll register the class
  directly, with the necessary side effect of initialize Objective-C
  metadata.
* @NSKeyedArchiveSubclassesOnly promises not to archive the class
  directly anyway.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
I believe I've addressed all of the feedback *except* for computing the Swift 3 mangled name.

DougGregor commented:
@swift-ci please smoke test

eeckstein commented:
@DougGregor You can get the old mangled name by demangling and re-mangling with the old remangler (which I will re-introduce shortly). We can talk about the details offline.

jrose-apple commented on include/swift/Serialization/ModuleFormat.h:
```diff
@@ -1314,6 +1314,8 @@ namespace decls_block {
     = BCRecordLayout<SynthesizedProtocol_DECL_ATTR>;
   using ImplementsDeclAttrLayout = BCRecordLayout<Implements_DECL_ATTR>;
+  using NSKeyedArchiveLegacyDeclAttrLayout
+    = BCRecordLayout<NSKeyedArchiveLegacy_DECL_ATTR>;
```
No longer necessary now that it's not serialized.

DougGregor replied:
We need the stub there because of various bits of preprocessor metaprogramming; note that the attribute above it (`ImplementsAttr`, `SynthesizedProtocolAttr`) are also not serialized.

jrose-apple replied:
Whoops, okay. Hm.

jrose-apple reviewed (commented):
🎉
